source/common/protobuf/utility.cc

```diff
@@ -596,6 +599,118 @@ std::string MessageUtil::CodeEnumToString(ProtobufUtil::error::Code code) {
   }
 }
 
+namespace {
+
+std::unique_ptr<Protobuf::Message> typeUrlToNewMessage(Protobuf::MessageFactory& message_factory,
+                                                       absl::string_view type_url) {
+  const absl::string_view type_name = TypeUtil::typeUrlToDescriptorFullName(type_url);
+  const Protobuf::Descriptor* descriptor =
+      Protobuf::DescriptorPool::generated_pool()->FindMessageTypeByName(
+          static_cast<std::string>(type_name));
+  return std::unique_ptr<Protobuf::Message>(
+      descriptor == nullptr ? nullptr : message_factory.GetPrototype(descriptor)->New());
+}
+
+// Forward declaration for mutually-recursive helper functions.
+void redact(Protobuf::Message* message, bool ancestor_is_sensitive);
+
+// To redact an `Any`, we have to unpack it to its original type to redact it.
+bool redactAny(Protobuf::Message* message, bool ancestor_is_sensitive) {
+  auto* any = dynamic_cast<ProtobufWkt::Any*>(message);
+  if (any != nullptr) {
+    Protobuf::DynamicMessageFactory message_factory;
+    auto typed_message = typeUrlToNewMessage(message_factory, any->type_url());
+    if (typed_message == nullptr) {
+      // If the type URL doesn't correspond to a known proto, give up redacting and treat this
+      // message the same as a `ProtobufWkt::Struct`. See the documented limitation on
+      // `MessageUtil::redact()` for more context.
+      ENVOY_LOG_MISC(warn, "Could not redact ProtobufWkt::Any with unknown type URL {}",
```

Comment:
Are there tests that check for this condition (ideally doing an `EXPECT_THROW_WITH_MESSAGE` on this message)?

Reply:
There are tests for this condition, although not using `EXPECT_THROW_WITH_MESSAGE`. We don't throw here, we fall back to treating the input as a struct.